lib/groups.js

```diff
@@ -191,38 +332,148 @@ module.exports.createGroup = function (groupName, assessmentId, userId, authnUse
+              if (notRequired && requiredGroupRoleIds.length > 0) {
+                return {
+                  user_id,
+                  group_role_id: requiredGroupRoleIds.pop(),
+                  group_id: groupId,
+                };
+              } else {
+                return { user_id, group_role_id, group_id: groupId };
+              }
+            }
+          );
+
+          // Assign remaining required roles to another group user
+          const assigneeUser = groupInfo.groupMembers.find((m) => m.user_id !== userId.toString());
+          for (const groupRoleId of requiredGroupRoleIds) {
+            groupRoleAssignmentUpdates.push({
+              user_id: assigneeUser.user_id,
+              group_role_id: groupRoleId,
+              group_id: groupId,
+            });
+          }
+
+          await sqldb.queryAsync(sql.clear_group_roles, { group_id: groupId });
+          await sqldb.queryAsync(sql.reassign_group_roles_after_leave, {
+            role_assignments: JSON.stringify(groupRoleAssignmentUpdates),
+          });
+        }
+
+        // Groups with low enough size should only use required roles
+        if (currentSize - 1 <= requiredRoles.length) {
+          await sqldb.queryAsync(sql.delete_non_required_roles, {
+            group_id: groupId,
+            assessment_id: assessmentId,
+          });
+        }
+      }
+    }
+
+    // Delete user from group and log
+    await sqldb.queryAsync(sql.delete_group_users, {
+      group_id: groupId,
+      user_id: userId,
+      authn_user_id: authnUserId,
+    });
   });
 };
 
+/**
+ * @param {string} assessmentId
+ * @param {string} userId
+ * @returns {Promise<AssessmentLevelPermissions>}
+ */
 module.exports.getAssessmentLevelPermissions = async function (assessmentId, userId) {
   const params = { assessment_id: assessmentId, user_id: userId };
-  const result = await sqldb.queryAsync(sql.get_assessment_level_permissions, params);
-  return result.rows[0];
+  return await sqldb.queryValidatedOneRow(
+    sql.get_assessment_level_permissions,
+    params,
+    AssessmentLevelPermissionsSchema
+  );
 };
 
+/**
+ * Updates the role assignments of users in a group, given the output from groupRoleSelectTable.ejs.
+ *
+ * @param {Object} requestBody
+ * @param {string} assessmentId
+ * @param {string} userId
+ * @param {string} authnUserId
+ */
 module.exports.updateGroupRoles = async function (requestBody, assessmentId, userId, authnUserId) {
-  // Convert form data to valid input format for a SQL function
-  const roleKeys = Object.keys(requestBody).filter((key) => key.startsWith('user_role_'));
-  const userIdToRoleIdMap = {};
-  for (const roleKey of roleKeys) {
-    const [roleId, userId] = roleKey.replace('user_role_', '').split('-');
-    if (userIdToRoleIdMap[userId] === undefined) {
-      userIdToRoleIdMap[userId] = [roleId];
-    } else {
-      userIdToRoleIdMap[userId].push(roleId);
+  await sqldb.runInTransactionAsync(async () => {
```

Comment:
@mwest1066 what do you think - should we be explicitly locking the `group_configs` row when doing all this?